File: 프로그래머스/lv2/12981. 영어 끝말잇기/영어 끝말잇기.py
def solution(n, words):
    answer = []
    appeared = [words[0]]

    for i in range(1, len(words)):
        # print(words[i])
        # print(appeared)
        # 탈락 조건은 나왔던 단어 말하거나, 끝말잇기 조건 위배한 경우
        if words[i] in appeared or words[i][0] != words[i-1][-1]:            
            answer = [(i % n) + 1, (i // n) +1]
            return answer
        
        # 나왔던 단어 목록에 추가하기
        appeared.append(words[i])
        

    return [0, 0]

File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.java
import java.util.*;

class Solution {
    public int[] solution(int n, String[] words) {
        int idx = 0;
        HashSet<String> checkedWord = new HashSet<>();
        char firstChar = ' ';
        
        while(idx < words.length){             
            String curWord = words[idx];
            System.out.println(idx + ", curWord: " + curWord);
            // 처음 등장 단어면 통과
            if(idx == 0) {
                firstChar = curWord.charAt(curWord.length() - 1);
                checkedWord.add(curWord);
                idx++;
                continue;
            }
            
            // 다음 단어로 넘어가기 전 검사
            if(curWord.charAt(0) != firstChar || checkedWord.contains(curWord)) {
                System.out.println("defeated! -> " + idx + ", curWord: " + curWord);
                int defeatedUserIdx = (idx) % n + 1;
                int defeatedTurnIdx = idx / n + 1;
                return new int[] {defeatedUserIdx, defeatedTurnIdx};
            }
            
            idx++;                        
            
            firstChar = curWord.charAt(curWord.length() - 1);
            checkedWord.add(curWord);                        
        }

        return new int[] {0, 0};
    }
}